keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/binary/AdditionVertex.java

```diff
@@ -18,19 +11,9 @@
      * @param right a vertex to add
      */
     public AdditionVertex(DoubleVertex left, DoubleVertex right) {
-        super(checkHasSingleNonScalarShapeOrAllScalar(left.getShape(), right.getShape()), left, right);
-    }
-
-    @Override
-    public DualNumber calculateDualNumber(Map<Vertex, DualNumber> dualNumbers) {
-        DualNumber leftDual = dualNumbers.get(left);
-        DualNumber rightDual = dualNumbers.get(right);
-        return leftDual.plus(rightDual);
+        super(left, right,
```

Comment:
Converting the op and dualNumber calculation to a lambda adds complexity IMHO and increases the memory footprint. I don't see us getting much out of it other than removing the *OpLambdas and removing a couple of boilerplate lines. I do really like removing the calculateDualNumbers on the ops up to the op and then calling a method defined by the op implementation that just requires it's left and right dual number instead of the entire world of dual numbers. I also like but am not sure about having the default shape check be an element-wise operations shape check. On one hand it is definitely by far the most common case and removing the boilerplate is nice but on the other hand if you don't require a shape from the op then someone creating a new op might not know that this shape constraint has been introduced. 

Can we keep the abstract op method and create an abstract dualNumberOp method that represents these two lambdas? Also, lets keep the default shaping constructor (good call) as I think other shape checks are uncommon (i.e. matrix multiply).

Reply:
Sure, I'm not too opposed to using an abstract protected method instead of a lambda function field.